Escreve seção explicação problema de CP (#20)
Rockett12/GEMP-UFPA: roteiro-de-estudos/cp-problem-explanation.md

  * [Solucionando o Problema de Programação Competitiva](#solucionando-o-problema-de-programação-competitiva)
  * [Submetendo a Solução](#submetendo-a-solução)
  * [O Juiz](#o-juiz)
  * [Vereditos](#vereditos)

## O que é Programação Competitiva?

Inicialmente, é necessário que não restem dúvidas para o leitor acerca das definições que estão sendo empregadas. Dessa forma, é possível descrever Programação Competitiva a partir da seguinte definição de um experiente programador competitivo [Ashar Fuadi](https://www.quora.com/What-is-competitive-programming-2/answer/Ashar-Fuadi)

> *“Competitive programming is solving well-defined problems by writing computer programs under specified limits”*

A definição acima pode ser livremente traduzida para o Português da seguinte maneira

> *“Programação Competitiva trata-se da resolução de problemas bem definidos mediante a escrita de programas de computador dentro de limites especificados"*

A partir da definição de Programação Competitiva, é possível ainda descrever com mais detalhes alguns termos presentes na definição, como *problemas bem definidos* e *programas de computador dentro de limites especificados*.

> O termo *problemas bem definidos*, indica que cada problema de Programação Competitiva possuirá as seguintes características: Descrição do ponto de partida do programa (informações e estados do qual a execução do programa partirá), definição das regras e das operações permitidas na situação problema e uma definição objetiva do estado final do problema (isto é, de sua solução). Por conta dessas características, existirá somente um único conjunto de respostas correspondente ao que é pedido no problema. Uma outra forma de entender o trecho é que um problema de programação competitiva não abre margem para o leitor interpretar de maneira diferente quais seriam as características (ponto de partida, regras e estado final) do problema que são consideradas aceitáveis.

Reviewer: Esse parágrafo ainda está um pouco confuso, com palavras complicadas pra um leitor leigo.

Uma sugestão é que seja mais direto, como em:
Uma problema bem definido vai te dizer exatamente como vai ser a entrada pro seu programa, como vai ser a saída, e o que você pode assumir sobre as informações contidas no enunciado (restrições sobre as variáveis envolvidas). Não dando margem para ambiguidades.

Author: Certo, eu vou adicionar um texto bem semelhante a sua sugestão, confesso que não conseguiria pensar em nada mais objetivo de imediato.

Reviewer 2: Concordo com o @Rockett12

Quanto ao termo *programa de computador dentro de limites especificados*, este indica que o programa de computador desenvolvido para solucionar o problema não pode ser feito de qualquer forma, existem limites impostos pelo problema que devem ser respeitados, em geral os limites dizem respeito ao tempo de execução do código, o valor que os parâmetros do problema podem assumir e o tamanho de memória que o programa pode alocar no máximo, caso os limites do problema sejam ultrapassados ou não respeitados devidamente, então a solução não será considerada aceitável.

Após uma explicação mais detalhada da definição e de alguns trechos que fazem parte desta, pode ser que o leitor se veja confuso e na expectativa de exemplos de como essas definições se apresentam na prática. Dessa maneira, a próxima seção lidará diretamente com a explicação de um problema de Programação Competitiva.   

## Como funciona um problema de Programação Competitiva?

A imagem a seguir contém a descrição do problema [Média Inteira](https://neps.academy/br/exercise/136) retirado da Plataforma NEPS Academy.

<p align="center">
  <img src="/images/media-inteira-neps.png">
</p>

> A partir da imagem, é possível separar o texto do problema em cinco principais seções: **Descrição do problema**, **Entrada e saída**, **Restrições** ,**Exemplos de teste** e **Limites**. É importante ressaltar que as cinco seções descritas sempre estarão presentes num problema de programação competitiva, pois são estas seções que garantem a definição de um *problema bem definido*.

Reviewer: Só faltou ajeitar a pontuação ali depois de **Restrições**

A seguir têm-se uma definição mais elaborada de cada uma das cinco seções destacadas

* **Descrição do Problema**: Trata-se de uma seção inicial em que geralmente o escritor do problema define um contexto para então apresentar uma situação-problema que deve ser resolvida pelo programador. No entanto, em alguns casos (como no exemplo em questão) não existe uma contextualização da situação problema e a tarefa é proposta para o programador de maneira direta, sem que seja necessária uma interpretação do mesmo sob o texto. No exemplo em questão, o problema pede para o programador calcular a média inteira entre dois valores inteiros denotados por **A** e **B**. 

> * **Entradas e Saídas**: Após a descrição do problema e do que o programador deve solucionar, a seção de entradas e saídas indica como devem ser inseridas e mostradas as informações que farão parte do problema. A **Entrada** diz respeito aos valores que serão inseridos no programa pelo teclado (ou por outros meios) para serem processados, quanto aos valores de resposta para o problema denotamos como **Saída**. Como dito, a seção de entradas e saídas indica EXATAMENTE como as informações correspondentes devem ser recebidas e mostradas pelo programa de computador, **caso as instruções dessa seção não sejam devidamente respeitadas, a solução não será considerada válida**.

Reviewer: Aquele "quanto" em 

> para serem processados, quanto aos

Não está correto.
Sugiro essa troca:
```suggestion
* **Entradas e Saídas**: Após a descrição do problema e do que o programador deve solucionar, a seção de entradas e saídas indica como devem ser inseridas e mostradas as informações que farão parte do problema. A **Entrada** diz respeito aos valores que serão inseridos no programa pelo teclado (ou por outros meios) para serem processados, E a **Saída** a como as respostas para o problema devem ser mostradas na tela. Como dito, a seção de entradas e saídas indica EXATAMENTE como as informações correspondentes devem ser recebidas e mostradas pelo programa de computador, **caso as instruções dessa seção não sejam devidamente respeitadas, a solução não será considerada válida**.
```

* **Restrições**: A seção de restrições fornece uma descrição dos valores que os parâmetros do problema assumirão durante a correção da solução do programador. No exemplo em questão, a restrição indica que os valores inseridos na entrada e utilizados durante o processamento do programa não possuirão um valor absoluto maior que 1000, isto é, o programa com a solução será avaliado somente considerando casos em que os valores inseridos respeitam essa restrição. Dessa forma, se por acaso o programa não conseguir gerar médias inteiras com valores de entrada maiores que 1000, a resposta continuará sendo considerada como correta, pois a solução exige um cálculo de médias inteiras apenas pra números na entrada cujo valor absoluto seja menor ou igual a 1000.

> * **Exemplos de Teste**: A seção fornece para o programador alguns exemplos de como o seu programa deve se comportar. Dessa maneira, são fornecidos alguns valores de entrada e saída, **respeitando as condições impostas na seção de Entrada e Saída**, e a lógica necessária para resolver o problema. Uma prática interessante é entender por que os casos de exemplo retornam determinada saída para uma dada entrada, pois assim o programador tem a confirmação de que este realmente entendeu o que o problema pede que seja solucionado. No exemplo em questão, o primeiro caso de teste conta com dois números inteiros de entrada (**2** e **2**) e um inteiro **2** na saída, isso indica que a operação de média inteira realizada entre os valores da entrada terá como resultado na saída o número **2**. Por fim, é importante ressaltar que os casos de teste pertencentes a seção não serão os únicos considerados durante a checagem da soluço desenvolvida pelo programador, sendo assim, **resolver os casos de teste da seção de exemplos não é uma garantia de que sua solução está correta**.

Reviewer 2: Typo
#
```suggestion
* **Exemplos de Teste**: A seção fornece para o programador alguns exemplos de como o seu programa deve se comportar. Dessa maneira, são fornecidos alguns valores de entrada e saída, **respeitando as condições impostas na seção de Entrada e Saída**, e a lógica necessária para resolver o problema. Uma prática interessante é entender por que os casos de exemplo retornam determinada saída para uma dada entrada, pois assim o programador tem a confirmação de que este realmente entendeu o que o problema pede que seja solucionado. No exemplo em questão, o primeiro caso de teste conta com dois números inteiros de entrada (**2** e **2**) e um inteiro **2** na saída, isso indica que a operação de média inteira realizada entre os valores da entrada terá como resultado na saída o número **2**. Por fim, é importante ressaltar que os casos de teste pertencentes a seção não serão os únicos considerados durante a checagem da solução desenvolvida pelo programador, sendo assim, **resolver os casos de teste da seção de exemplos não é uma garantia de que sua solução está correta**.
```

> * **Limites**: Os limites impostos pelo problema dizem respeito ao tempo máximo de execução que o programa de solução deve levar para resolver um caso de teste e a quantidade de memória máxima que o programa irá alocar no computador para resolver cada caso de teste. De maneira resumida, um programa de computador pode demorar menos ou mais tempo para processar uma determinada quantidade de informações de acordo com as características dessa informação. Além disso, cada informação recebida ou gerada por um programa de computador aloca uma quantidade definida de espaço na memória que varia de acordo com as características da informação, sendo assim é possível ter soluções diferentes que consumam espaços diferentes na memória. No exemplo em tese, o tempo limite de execução do código é de 1 segundo e o espaço máximo que pode ser ocupado é de 256 megabytes (MB), **caso uma solução não respeite os limites de memória e tempo do problema, ela não será considerada aceitável**.

Reviewer 2: Removi algumas palavras que julguei desnecessárias, e modifiquei a estrutura do texto.

```suggestion
* **Limites**: Os limites impostos pelo problema dizem respeito ao tempo máximo de execução que o programa deve levar para resolver um caso de teste e a quantidade de memória máxima que o programa irá alocar no computador para resolver cada caso de teste. De maneira resumida, um programa de computador pode demorar menos ou mais tempo para processar uma determinada quantidade de informações, assim como a quantidade de espaço alocada na memória pode variar, de acordo com a lógica do programa. No exemplo em tese, o tempo limite de execução do código é de 1 segundo e o espaço máximo que pode ser ocupado é de 256 megabytes (MB), **caso uma solução não respeite os limites de memória e tempo do problema, ela não será considerada aceitável**.
```
